### PR TITLE
FEATURE: Delegated authentication via user api keys

### PR DESCRIPTION
**Pull request by pmusaraj**
For details, see: https://meta.discourse.org/t/delegated-authentication-for-discourse-mobile-app/110414

This adds a new capability to the user API: delegated authentication via a `one time password` mechanism. This allows apps (like the DiscourseHub mobile app) to request a one-use password and use that to authenticate the user in an isolated browser session (like SafariViewController, for example).

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -359,6 +360,25 @@ def set_mobile_view
     session[:mobile_view] = params[:mobile_view] if params.has_key?(:mobile_view)
   end
 
+  def handle_safari_svc_session
+    # Workaround for dropped cookies in DiscourseHub SafariViewController iOS 12+.
```
Comment (SamSaffron) [inline app/controllers/application_controller.rb]: do we want to call this "safari" specific I agree safari is the symptom but really any client could opt for the same magic no?

Comment (pmusaraj) [reply app/controllers/application_controller.rb]: You're right, of course, no need to call this Safari anything. My initial hunch was to explain the use case for this now, but it's already usable by any client. (Which is also why there is an `auth_redirect` parameter, to let clients use their own URL here.)

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -46,6 +46,7 @@ def handle_unverified_request
   before_action :clear_notifications
   before_action :set_locale
   before_action :set_mobile_view
+  before_action :handle_delegated_auth
```
Comment (SamSaffron) [inline app/controllers/application_controller.rb]: can we avoid adding an extra `before action` I am pretty protective of having too many ? maybe just do this direct in current user provider ?

File: config/site_settings.yml [hunk 3]
```diff
@@ -1919,7 +1919,7 @@ user_api:
   allow_user_api_keys:
     default: true
   allow_user_api_key_scopes:
-    default: "read|write|message_bus|push|notifications|session_info"
+    default: "read|write|message_bus|push|notifications|session_info|one_time_password"
```
Comment (SamSaffron) [inline config/site_settings.yml]: what is the transition period going to look like now that existing keys kind of want to have one_time_password?

**Comment by SamSaffron** [comment]
Overall I am very happy with this change, feel free to merge once you address my small niggles

File: config/site_settings.yml (hunk 3 above)
Comment (pmusaraj) [reply config/site_settings.yml]: It's pretty straightforward: sites without the one_time_password scope won't get the redirect treatment. This is valid for both sites that aren't up to date as well as sites that decide to disallow that scope. 

The PR also bumps the user API version to 4, so apps can check whether `one_time_password` is supported by checking the API version.

**Pull request merged**